Deprecates X-XSS-Protection and put it behind feature flag.

akbar1214 opened the pull request:
Closes #21728
- Marks X-XSS-Protection  header as deprecated.
- Put X-XSS-Protection behind a feature flag.
- Add/Update relavant test case.

akbar1214 committed:
fix/deprecate-xss-header:

- Put X_XSS_PROTECTION header behind a feature flag.
- Removes it from defaultHeaders.

Signed-off-by: akbarhusainpatel <apatel@intermiles.com>

Commit:
fix/deprecate-xss-header:

- Updates existing AdminHeadersTest to validate header is not set by default.
- Adds AdminHeadersXSSHeaderEnabledTest to validate header is set if feature flag is enabled.

Signed-off-by: akbarhusainpatel <apatel@intermiles.com>

ahus1 commented:
Please update your PR description with `Closes #xxxx` so it references and open issue that it is about to fix.

akbar1214 committed:
fix/deprecate-xss-header:

- Converts im-mutable map to mutable during modification.

Signed-off-by: akbarhusainpatel <apatel@intermiles.com>

akbar1214 commented:
@jonkoops can someone review the pr. Out of the 5 failure 4 are realted to database image pull from registry. And the other one is Keycloak CI Status Check. None of these failure seems realted to my PR.

ahus1 commented:
I've restarted the failed builds.

jonkoops commented on server-spi-private/src/main/java/org/keycloak/models/BrowserSecurityHeaders.java:
```diff
@@ -65,7 +66,6 @@ public String getDefaultValue() {
         dh.put(CONTENT_SECURITY_POLICY_REPORT_ONLY.getKey(), CONTENT_SECURITY_POLICY_REPORT_ONLY.getDefaultValue());
         dh.put(X_CONTENT_TYPE_OPTIONS.getKey(), X_CONTENT_TYPE_OPTIONS.getDefaultValue());
         dh.put(X_ROBOTS_TAG.getKey(), X_ROBOTS_TAG.getDefaultValue());
-        dh.put(X_XSS_PROTECTION.getKey(), X_XSS_PROTECTION.getDefaultValue());
```
Is there any reason why we would not be able use `Profile.isFeatureEnabled()` here?

akbar1214 replied:
Since static blocks are called during class loading Profile.getInstance() will return null. As the CURRENT field is initialised when Profile#init method is called.

jonkoops replied:
Figured as much, then indeed we have no choice but to repeat this condition, which considering this is a deprecation is fine to me as the code will be removed eventually.

akbar1214 replied:
Since this task will deprecate the header should we create new issue for removal as well ?

jonkoops reviewed (commented):
Largely looks good to me, just some small nits.

akbar1214 committed:
Update metadata for  BrowserSecurityHeaders.java

Co-authored-by: Jon Koops <jonkoops@gmail.com>
Signed-off-by: akbar1214 <43771058+akbar1214@users.noreply.github.com>

jonkoops reviewed (dismissed):
LGTM, I will get a maintainer to review this as well.

stianst reviewed (commented):
I think adding this behind a feature is way overboard, and not the right use of features. From a feature perspective the feature is security headers, not individual headers.

Additionally if this is considered a breaking change then strictly speaking it should still be enabled by default.

I reckon we can just remove this header without wrapping this in a option. Assuming that the statement is correct and the header is no longer relevant to be set.

Alternatively, just make it a config option on the provider instead.

stianst reviewed (dismissed):
See previous comment

akbar1214 commented:
@jonkoops IMHO we should remove the header since the current implementation also removes it by default. Also all tests are green. 

Let me know if you have different thoughts.

jonkoops commented:
I agree with you and Stian, let's remove this header entirely, it isn't actually honored by any user agents we actively support anyways. Please do add a section to the changelog for `26.2.x` that we have removed this in case anyone still relies on it.

Commit:
fix/deprecate-xss-header:

- Removes X-XSS-Protection header completely.
- Update test case to validate that header is not set.

Signed-off-by: akbarhusainpatel <apatel@intermiles.com>

akbar1214 commented:
@jonkoops / @stianst can you please review the PR. For the failing Javascript CI I believe it is related to #36998

jonkoops commented on model/storage-private/src/main/java/org/keycloak/storage/datastore/DefaultExportImportManager.java:
```diff
@@ -63,7 +63,6 @@
 import org.keycloak.models.utils.DefaultKeyProviders;
 import org.keycloak.models.utils.DefaultRequiredActions;
 import org.keycloak.models.utils.KeycloakModelUtils;
-import org.keycloak.models.utils.ModelToRepresentation;
```
Let's try to keep unrelated code changes out of the PR.

jonkoops commented on tests/base/src/test/java/org/keycloak/tests/admin/AdminHeadersTest.java:
I don't think we have to write a test to ensure functionality doesn't exist, these changes can be removed.

jonkoops reviewed (changes requested):
We should also remove the [ability to set the value](https://github.com/keycloak/keycloak/blob/5ea7825f1e962ec971d2eef348d444c81c2ee73a/js/apps/admin-ui/src/realm-settings/security-defences/HeadersForm.tsx#L51-L54) of this header from the admin console. Also, if you search for the name of the field as stored you will find `xXSSProtection` referenced in a lot of places where it can be removed.

Commit:
fix/deprecate-xss-header:

- Updates documentation.
- Remove XSS Header.

Signed-off-by: akbarhusainpatel <apatel@intermiles.com>

jonkoops reviewed (commented):
Very thorough, looks great! I just want to run the copy by our docs team and then I think we can get this merged.

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -0,0 +1,7 @@
+We have removed the X-XSS-Protection response header from Keycloak, this header was a feature of Internet Explorer,
+Chrome and Safari that stopped pages from loading when they detected reflected cross-site scripting (XSS) attacks.
+
+Since, this header is no longer supported in any of the user agents supported by Keycloak and has been replaced by
```
```suggestion
Since this header is no longer supported in any user agents that are supported by Keycloak and has been replaced by
```

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -0,0 +1,7 @@
+We have removed the X-XSS-Protection response header from Keycloak, this header was a feature of Internet Explorer,
+Chrome and Safari that stopped pages from loading when they detected reflected cross-site scripting (XSS) attacks.
+
+Since, this header is no longer supported in any of the user agents supported by Keycloak and has been replaced by
+Content-Security-Policy. More details about the header can be found https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-XSS-Protection[here]
```
```suggestion
Since this header is no longer supported in any of the user agents supported by Keycloak, it has been replaced by
Content-Security-Policy. For more details, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-XSS-Protection[X-XSS-Protection].
```

andymunro reviewed (dismissed):
A few comments, @jonkoops

jonkoops committed:
Update docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc

Signed-off-by: Jon Koops <jonkoops@gmail.com>

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -0,0 +1,5 @@
+= Removal of the `X-XSS-Protection` header
+
+Because the https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-XSS-Protection[`X-XSS-Protection` header] is no longer supported by any of the user agents that are supported by Keycloak it has been removed. This header was a feature of Internet Explorer, Chrome and Safari that stopped pages from loading when they detected reflected cross-site scripting (XSS) attacks.
```
```suggestion
Because the https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-XSS-Protection[`X-XSS-Protection` header] is no longer supported by any user agents that are supported by Keycloak, it has been removed. This header was a feature of Internet Explorer, Chrome, and Safari that stopped pages from loading when they detected reflected cross-site scripting (XSS) attacks.
```

jonkoops committed:
Update docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Jon Koops <jonkoops@gmail.com>

jonkoops reviewed (dismissed):
LGTM!

rmartinc reviewed (approved):
Just one comment. The current attribute `_browser_header.xXSSProtection` is not removed from the migrated realms. It will be a leftover in the realms that is now present. This is not breaking anything though. I see there is no `MigrateTo26_2_0` class yet, so I'm not saying to do it only for this change and I'm approving the PR.

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/resources/migration-test/migration-realm-13.0.1-client-policies.json:
```diff
@@ -1975,7 +1975,6 @@
     "xRobotsTag" : "none",
     "xFrameOptions" : "SAMEORIGIN",
     "contentSecurityPolicy" : "frame-src 'self'; frame-ancestors 'self'; object-src 'none';",
-    "xXSSProtection" : "1; mode=block",
```
Is it possible to revert the change in all the `migration-realm-*.json` files and make sure that this line with `xXSSProtection` is still present?

The point of those migration tests is to test the scenarios like:
- Exporting realm in Keycloak 13.0.1 (or any other Keycloak version)
- Then make sure that exported JSON file from previous step can be successfully imported into the latest Keycloak

The JSON files exported in production in Keycloak 13.0.1 (or whatever other old version) will always have this line with `xXSSProtection` . So we need to make sure that it is possible to successfully import them in latest Keycloak server without any manual update needed in that JSON file. Hence it is not ideal to remove this line as this breaks the purpose of this test...

If things are broken when this line is removed, then some fix might be needed to make sure that it is still possible to import JSON files with this line present.

mposolda commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -0,0 +1,5 @@
+= Removal of the `X-XSS-Protection` header
```
Is it possible to add this to the release notes for 26.2.0 instead of upgrading guide? AFAIK upgrading guide may contain only things, which somehow affect backwards compatibility and require some action from the Keycloak administrators/users. But this is not the case if I understand correctly?

jonkoops replied:
This still might affect some users if they re-use this header, I think the changelog is the perfect place for it. Putting it in the release notes would just clutter them, I'd rather reserve it for a nice overview of features.

jonkoops replied:
Resolving this as per https://github.com/keycloak/keycloak/pull/36881#issuecomment-2666546172

mposolda reviewed (changes requested):
Added 2 comments inline. Also the rebase might be needed.

jonkoops commented:
@akbar1214 could you introduce a migration that removes the field from the database as @rmartinc mentioned?

Commit:
- Moves xss header removal related doc changes to release notes.

Signed-off-by: akbarhusainpatel <apatel@intermiles.com>

Commit:
- Reverts migration-realm-*.json related changes.

Signed-off-by: akbarhusainpatel <apatel@intermiles.com>

akbar1214 committed:
- Adds Migration for 26.2.0.

Signed-off-by: akbarhusainpatel <apatel@intermiles.com>

Commit:
- Removes xss attribute from realm.

Signed-off-by: akbarhusainpatel <apatel@intermiles.com>

mposolda reviewed (dismissed):
@akbar1214 Thanks for the updates, approving now.

@jonkoops The contributor added documentation to the release notes as I've mentioned. Are you ok with that?

If you think it can affect people migrating from previous versions, maybe it is worth to mention it on both release-notes and migration guide? In that case, upgrading-guide may just have small sentence like "We remove X-XSS protection header. For more details look at the release notes" or something like that. WDYT?

jonkoops commented:
I think it is fine to keep it in the release notes, we'll be going through them anyways before release so that gives us a chance to clean things up if needed. @akbar1214 could you rebase what you have now to fix the conflicts, I think we can merge this work after.

jonkoops commented:
Thanks a bunch for your contribution and patience @akbar1214!

akbar1214 commented:
Thanks for the support @jonkoops :)
